compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -393,15 +433,29 @@ pub fn intern_const_alloc_recursive<
     debug!(?todo);
     debug!("dead_alloc_map: {:#?}", ecx.memory.dead_alloc_map);
     while let Some(alloc_id) = todo.pop() {
-        if let Some((_, mut alloc)) = ecx.memory.alloc_map.remove(&alloc_id) {
+        if let Some((kind, mut alloc)) = ecx.memory.alloc_map.remove(&alloc_id) {
+            for &(_, alloc_id) in alloc.provenance().ptrs().iter() {
+                if leftover_allocations.insert(alloc_id) {
+                    todo.push(alloc_id);
+                }
+            }
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
             match intern_kind {
                 // Statics may point to mutable allocations.
                 // Even for immutable statics it would be ok to have mutable allocations behind
                 // raw pointers, e.g. for `static FOO: *const AtomicUsize = &AtomicUsize::new(42)`.
-                InternKind::Static(_) => {}
+                InternKind::Static(_) => {
+                    assert!(
+                        !matches!(kind, MemoryKind::StaticRoot(_)),
+                        "should have already gotten interned"
+                    );
+                    if let Some(GlobalAlloc::Static(_)) = ecx.tcx.try_get_global_alloc(alloc_id) {
+                        ecx.memory.alloc_map.insert(alloc_id, (kind, alloc));
```

Comment:
What is this doing and why? The only allocation we should have to re-insert is the root allocation.

Reply:
yea, but the root allocation may be referred to by the static itself (`static FOO: Thing = Thing {`), and so during the

Reply:
the leftover allocations are also removed from the `alloc_map` just a few lines above this line that you commented on.

Reply:
Ah lol so we're removing and adding and re-removing and re-adding... :joy: